core/corehttp/gateway_handler.go

```diff
@@ -216,8 +227,32 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	}
 
 	parsedPath := ipath.New(urlPath)
-	if err := parsedPath.IsValid(); err != nil {
-		webError(w, "invalid ipfs path", err, http.StatusBadRequest)
+	if pathErr := parsedPath.IsValid(); pathErr != nil {
+		// Attempt to fix redundant /ipfs/ namespace as long as resulting
+		// 'intended' path is valid.  This is in case gremlins were tickled
+		// wrong way and user ended up at /ipfs/ipfs/{cid} or /ipfs/ipns/{id}
+		// like in bafybeien3m7mdn6imm425vc2s22erzyhbvk5n3ofzgikkhmdkh5cuqbpbq
+		// :^))
+		intendedPath := ipath.New(strings.TrimPrefix(urlPath, "/ipfs"))
```

Comment:
I'd slightly rather test these cases explicitly (`/ipfs/ipns`, `/ipfs/ipfs`, etc.). It's not critical, but it took me a few minutes to prove to myself that there was no way this could go wrong (e.g., can't redirect `/ipfs` to `/`, etc.).

(I don't feel strongly about this)

Reply:
Added explicit check in https://github.com/ipfs/go-ipfs/pull/7930/commits/450baef0e9c113e4751b559dd0562f7c98302668